File: src/main/java/com/ychstudio/gui/ConvertKitController.java
package com.ychstudio.gui;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.*;
import javafx.stage.FileChooser;
import opencc.OpenCC;
import org.mozilla.universalchardet.UniversalDetector;

import java.io.*;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.Map;
import java.util.Optional;
import java.util.ResourceBundle;

public class ConvertKitController implements Initializable {

    @FXML
    Label selectedFileLabel;

    @FXML
    RadioButton gbkRadioButton;

    @FXML
    RadioButton big5RadioButton;

    @FXML
    RadioButton utf8RadioButton;

    @FXML
    ComboBox<String> encodingComboBox;

    @FXML
    ComboBox<String> conversionComboBox;

    @FXML
    Button applyConversionButton;

    @FXML
    TextArea inputTextArea;

    @FXML
    TextArea outputTextArea;

    private Optional<File> selectedFile = Optional.empty();
    private OpenCC openCC = new OpenCC();

    EventHandler<ActionEvent> changeSourceEncodingHandler = event -> {
        if (event.getSource() == gbkRadioButton) {
            encodingComboBox.setValue("GBK");
        }
        else if (event.getSource() == big5RadioButton) {
            encodingComboBox.setValue("Big5");
        }
        else if (event.getSource() == utf8RadioButton) {
            encodingComboBox.setValue("UTF-8");
        }
        else if (event.getSource() == encodingComboBox) {
            switch (encodingComboBox.getValue()) {
                case "GBK":
                    gbkRadioButton.setSelected(true);
                    break;
                case "Big5":
                    big5RadioButton.setSelected(true);
                    break;
                case "UTF-8":
                    utf8RadioButton.setSelected(true);
                    break;
                default:
                    gbkRadioButton.setSelected(false);
                    big5RadioButton.setSelected(false);
                    utf8RadioButton.setSelected(false);
                    break;
            }
            changeInputEncoding(encodingComboBox.getValue());
        }

    };

    EventHandler<ActionEvent> conversionActionHandler = event -> {
        applyConversion();
    };

    EventHandler<ActionEvent> applyConversionHandler = event -> {
        applyConversion();
    };

    protected void applyConversion() {
        for (Map.Entry<String, String> entry : OpenCC.CONVERSIONS.entrySet()) {
            if (entry.getValue().equals(conversionComboBox.getValue())) {
                convertInputToOutput(entry.getKey());
                break;
            }
        }
    }

    public void changeInputEncoding(String encoding) {
        if (selectedFile.isPresent()) {
            loadFileWithEncoding(selectedFile.get(), encoding);
        }
    }

    public String detectFileEncoding(File file) {
        UniversalDetector universalDetector = new UniversalDetector(null);

        int nread;
        byte[] buffer = new byte[2048];

        try {
            BufferedInputStream bis = new BufferedInputStream(new FileInputStream(file));
            while((nread = bis.read(buffer)) > 0 && !universalDetector.isDone()) {
                universalDetector.handleData(buffer, 0, nread);
            }
            universalDetector.dataEnd();
        } catch (IOException e) {
            e.printStackTrace();
        }

        Optional<String> encoding = Optional.ofNullable(universalDetector.getDetectedCharset());
        return encoding.isPresent() ? encoding.get() : "UTF-8";
    }

    protected void loadFileWithEncoding(File file, String encoding) {
        StringBuilder stringBuilder = new StringBuilder();
        try {
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(new FileInputStream(file), encoding));
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                stringBuilder.append(line);
                stringBuilder.append("\n");
            }

            inputTextArea.clear();
            inputTextArea.setText(stringBuilder.toString());
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    protected void convertInputToOutput(String conversion) {
        String input = inputTextArea.getText();
        openCC.setConversion(conversion);
        String output = openCC.convert(input);
        outputTextArea.clear();
        outputTextArea.setText(output);
    }

    public void chooseFile() {

        FileChooser fileChooser = new FileChooser();
        selectedFile = Optional.ofNullable(fileChooser.showOpenDialog(null));
        if (selectedFile.isPresent()) {
            File file = selectedFile.get();
            selectedFileLabel.setText(file.getName());

            String encoding = detectFileEncoding(file);

            switch (encoding) {
                case "GBK":
                    gbkRadioButton.setSelected(true);
                    break;
                case "Big5":
                    big5RadioButton.setSelected(true);
                    break;
                case "UTF-8":
                    utf8RadioButton.setSelected(true);
                    break;
                default:
                    gbkRadioButton.setSelected(false);
                    big5RadioButton.setSelected(false);
                    utf8RadioButton.setSelected(false);
                    break;
            }
            encodingComboBox.setValue(encoding);

            loadFileWithEncoding(file, encoding);
        }
    }


    protected void buildConversionComboBox() {
        ObservableList<String> conversionOptions = FXCollections.observableArrayList();
        conversionComboBox.setItems(conversionOptions);

        OpenCC.CONVERSIONS.values().stream()
                .sorted()
                .forEach(conversionOptions::add);

    }


    @Override
    public void initialize(URL location, ResourceBundle resources) {
        ObservableList<String> encodingChoiceList = FXCollections.observableArrayList();
        encodingChoiceList.addAll(Charset.availableCharsets().keySet());
        encodingComboBox.setItems(encodingChoiceList);

        gbkRadioButton.setOnAction(changeSourceEncodingHandler);
        big5RadioButton.setOnAction(changeSourceEncodingHandler);
        utf8RadioButton.setOnAction(changeSourceEncodingHandler);
        encodingComboBox.setOnAction(changeSourceEncodingHandler);

        conversionComboBox.setOnAction(conversionActionHandler);
        buildConversionComboBox();

        applyConversionButton.setOnAction(applyConversionHandler);
    }
}

File: src/main/kotlin/com/ychstudio/gui/ConvertKitController.kt
package com.ychstudio.gui

import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.*
import javafx.stage.FileChooser
import opencc.OpenCC
import org.mozilla.universalchardet.UniversalDetector
import java.io.*
import java.net.URL
import java.nio.charset.Charset
import java.util.*

class ConvertKitController : Initializable {

    @FXML
    internal var quitMenuItem: MenuItem? = null

    @FXML
    internal var selectedFileLabel: Label? = null

    @FXML
    internal var gbkRadioButton: RadioButton? = null

    @FXML
    internal var big5RadioButton: RadioButton? = null

    @FXML
    internal var utf8RadioButton: RadioButton? = null

    @FXML
    internal var encodingComboBox: ComboBox<String>? = null

    @FXML
    internal var conversionComboBox: ComboBox<String>? = null

    @FXML
    internal var applyConversionButton: Button? = null

    @FXML
    internal var inputTextArea: TextArea? = null

    @FXML
    internal var outputTextArea: TextArea? = null

    @FXML
    internal var targetEncodingComboBox: ComboBox<String>? = null

    @FXML
    internal var saveAsRadioButton: RadioButton? = null

    @FXML
    internal var overwriteRadioButton: RadioButton? = null

    @FXML
    internal var saveFileButton: Button? = null


    private var selectedFile: File? = null
    private val openCC = OpenCC()

    internal var changeSourceEncodingHandler = { event: ActionEvent ->
        when (event.source) {
            gbkRadioButton -> encodingComboBox?.value = "GBK"
            big5RadioButton -> encodingComboBox?.value = "Big5"
            utf8RadioButton -> encodingComboBox?.value = "UTF-8"
            encodingComboBox -> {
                when (encodingComboBox?.value) {
                    "GBK" -> gbkRadioButton?.isSelected = true
                    "Big5" -> big5RadioButton?.isSelected = true
                    "UTF-8" -> utf8RadioButton?.isSelected = true
                    else -> {
                        gbkRadioButton?.isSelected = false
                        big5RadioButton?.isSelected = false
                        utf8RadioButton?.isSelected = false
                    }
                }
                changeInputEncoding(encodingComboBox!!.value)
            }
            else  -> {}
        }
    }

    internal var conversionActionHandler = { event: ActionEvent -> applyConversion() }

    internal var applyConversionHandler = { event: ActionEvent -> applyConversion() }

    protected fun applyConversion() {
        for (entry in OpenCC.CONVERSIONS.entries) {
            if (entry.value == conversionComboBox?.value) {
                convertInputToOutput(entry.key)
                saveFileButton?.isDisable = false
                break
            }
        }
    }

    fun changeInputEncoding(encoding: String) {
        if (selectedFile != null) {
            loadFileWithEncoding(selectedFile as File, encoding)
            applyConversion()
        }
    }

    fun detectFileEncoding(file: File): String {
        val universalDetector = UniversalDetector(null)

        val buffer = ByteArray(2048)

        try {
            val bis = BufferedInputStream(FileInputStream(file))
            var nread = bis.read(buffer)
            while (nread > 0 && !universalDetector.isDone) {
                universalDetector.handleData(buffer, 0, nread)
                nread = bis.read(buffer)
            }
            universalDetector.dataEnd()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        val encoding = universalDetector.detectedCharset
        return encoding ?: "UTF-8"
    }

    protected fun loadFileWithEncoding(file: File, encoding: String) {
        val stringBuilder = StringBuilder()
        try {
            val bufferedReader = BufferedReader(InputStreamReader(FileInputStream(file), encoding))
            var line: String? = bufferedReader.readLine()
            while (line != null) {
                stringBuilder.append(line)
                stringBuilder.append("\n")
                line = bufferedReader.readLine()
            }

            inputTextArea?.clear()
            inputTextArea?.text = stringBuilder.toString()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    protected fun convertInputToOutput(conversion: String) {
        val input = inputTextArea?.text
        openCC.setConversion(conversion)
        val output = openCC.convert(input)
        outputTextArea?.clear()
        outputTextArea?.text = output
    }

    fun chooseFile() {

        val fileChooser = FileChooser()
        selectedFile = fileChooser.showOpenDialog(null)
        if (selectedFile != null) {
            selectedFileLabel?.text = (selectedFile as File).name

            val encoding = detectFileEncoding(selectedFile as File)

            when (encoding) {
                "GBK" -> gbkRadioButton?.isSelected = true
                "Big5" -> big5RadioButton?.isSelected = true
                "UTF-8" -> utf8RadioButton?.isSelected = true
                else -> {
                    gbkRadioButton?.isSelected = false
                    big5RadioButton?.isSelected = false
                    utf8RadioButton?.isSelected = false
                }
            }
            encodingComboBox?.value = encoding

            loadFileWithEncoding(selectedFile as File, encoding)
            saveFileButton?.isDisable = true
            applyConversion()
        }
    }

    fun saveFile() {

        when {
            saveAsRadioButton!!.isSelected -> {
                val fileChooser = FileChooser()
                fileChooser.initialFileName = selectedFile!!.name
                fileChooser.initialDirectory = selectedFile!!.parentFile
                val chosen: File? = fileChooser.showSaveDialog(null)

                if (chosen != null) {
                    writeToFile(chosen.absolutePath, outputTextArea!!.text, targetEncodingComboBox!!.value)
                }
            }

            overwriteRadioButton!!.isSelected -> {
                writeToFile(selectedFile!!.absolutePath, outputTextArea!!.text, targetEncodingComboBox!!.value)
            }
        }
    }

    protected fun writeToFile(filename: String, output: String, encoding: String) {
        if (output.isEmpty()) {
            return
        }

        val outputFile = File(filename)

        val bufferedWriter= BufferedWriter(OutputStreamWriter(FileOutputStream(outputFile), encoding))
        bufferedWriter.write(output)
        bufferedWriter.flush()
        bufferedWriter.close()
    }

    protected fun buildConversionComboBox() {
        val conversionOptions = FXCollections.observableArrayList<String>()
        conversionComboBox?.items = conversionOptions

        val tmpArrayList = ArrayList<String>()
        for (value in OpenCC.CONVERSIONS.values) {
            tmpArrayList.add(value)
        }
        tmpArrayList.sort(String::compareTo)
        tmpArrayList.forEach { conversionOptions.add(it) }
    }


    override fun initialize(location: URL?, resources: ResourceBundle?) {

        quitMenuItem?.setOnAction({ e -> System.exit(0) })

        val encodingChoiceList = FXCollections.observableArrayList<String>()
        encodingChoiceList.addAll(Charset.availableCharsets().keys)
        encodingComboBox?.items = encodingChoiceList
        targetEncodingComboBox?.items = encodingChoiceList
        targetEncodingComboBox?.value = "UTF-8"

        gbkRadioButton?.setOnAction(changeSourceEncodingHandler)
        big5RadioButton?.setOnAction(changeSourceEncodingHandler)
        utf8RadioButton?.setOnAction(changeSourceEncodingHandler)
        encodingComboBox?.setOnAction(changeSourceEncodingHandler)

        conversionComboBox?.setOnAction(conversionActionHandler)
        buildConversionComboBox()

        applyConversionButton?.setOnAction(applyConversionHandler)

        inputTextArea?.isEditable = false
//        outputTextArea?.isEditable = false
    }
}
